woocommerce/woocommerce
# [CYS] Avoid duplicating requests during the CYS flow

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR refactors the `customizeStoreStateMachine` to avoid making the `/onboarding/themes/recommended` and `/onboarding/tasks`, request twice during the CYS flow:

- Sets the initial value for the `activeThemeHasMods` variable to `undefined` to mark it as "not fetched".
- Extracts the `fetchActiveThemeHasMods` function from the `fetchIntroData` since it's not needed in the `intro` step. 
- Add a first `setFlags` step that fetches some values to be available in the context.
- The `fetchActiveThemeHasMods` is removed from the `assemblerHub` first step since the variable is on the context.
- Sets the parent window `activeThemeHasMods` to true after the `designWithAI` and  `designWithoutAI` steps.

Part of #43856 



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Make sure to `Reset Customize Your Store` on `wp-admin/tools.php?page=woocommerce-admin-test-helper` if the store is not newly created.
2. Go to `WooCommerce > Home`.
3. Open the network tab in your browser.
4. Click on the `Start Customizing` button.
5. Check the `/onboarding/themes/recommended` and `/onboarding/tasks` requests are only done once (before this PR they were requested twice).

**Check the redirection to intro works**

0. Make sure to `Reset Customize Your Store` on `wp-admin/tools.php?page=woocommerce-admin-test-helper` if the store is not newly created.
1. Go to `wp-admin/admin.php?page=wc-admin&path=/customize-store/assembler-hub`.
2. Check it redirects to the intro screen (should not show the assembler).
3. Go to `WooCommerce > Home`, click on the `Start Customizing` button and finish the flow.
4. Go to `wp-admin/admin.php?page=wc-admin&path=/customize-store/assembler-hub`.
5. Check this time it does not redirect (should show the assembler).
6. Repeat 1-5 in a WooExpress store.



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Avoid duplicating requests to the `/onboarding/themes/recommended` and `/onboarding/tasks` endpoints during the CYS flow.

#### Comment 

</details>
